### PR TITLE
Refactorisation of cost function and weighting

### PR DESCRIPTION
**Pull request by fiqas**
To make my life easier with all that weighting that I do, I've made cost and weighting two separate modules - loss.h/cpp and weights.h/cpp.

I have not added my weighting scheme, just Roman's data-weighting.
The code was only refactorised. If my experiments succeed, I will add my weighting scheme in the future. If anybody wants to add any other type of weighting, they can easily do so by overriding weighting class base.

Here are the regression tests results. The same stuff fails in clean marian-dev without my changes.

screen shot 2018-04-26 at 15 38 18

### CODE REVIEW

**Comment by fiqas** [comment]
Re-run the tests with downloaded models:

Skipped:
  - tests/decoder/wmt16/test_ende_cpu.sh
  - tests/examples/iris/test_iris.sh
  - tests/examples/mnist/test_mnist_ffnn.sh
  - tests/examples/mnist/test_mnist_lenet.sh
  - tests/server/test_ende_batch32.sh
  - tests/server/test_ende.sh
  - tests/models/char-s2s/test_char_s2s.sh
  - tests/models/char-s2s/test_compare_scores.sh
  - tests/training/restart-optimizer/test_adam_params_async.sh
  - tests/training/restart-optimizer/test_adam_params_sync.sh
  - tests/training/multi-gpu/test_async_sgd.sh
Failed:
  - tests/training/restart/test_restarting_stalled_validators.sh
  - tests/training/restart-optimizer/test_loading_adam_params.sh

**Comment by emjotde** [comment]
Waiting with merge until you can demonstrate this works :)

**Comment by fiqas** [comment]
This doesn’t implement my weighting — it’s just refactoring :) When it comes to weighting, there is only data-weighting., just as it was.

**Comment by emjotde** [comment]
Ah right. Them I am reassigning this to @snukky for review.

**Comment by snukky** [comment]
Merged and I added some regression tests for different loss functions.
